modules/bm_sim/include/bm_sim/match_key_types.h

```diff
@@ -0,0 +1,78 @@
+/* Copyright 2013-present Barefoot Networks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Gordon Bailey (gb@gordonbailey.net)
+ *
+ */
+
+#ifndef BM_SIM_INCLUDE_BM_SIM_MATCH_KEY_TYPES_H_
+#define BM_SIM_INCLUDE_BM_SIM_MATCH_KEY_TYPES_H_
+
+#include <limits>
+
+#include "bytecontainer.h"
+
+namespace bm {
+
+typedef uintptr_t internal_handle_t;
+typedef uint64_t entry_handle_t;
+
+enum class MatchUnitType {
+  EXACT, LPM, TERNARY
+};
+
+// Entry types.
+struct MatchKey {
+  MatchKey() {}
+  MatchKey(ByteContainer data, uint32_t version)
+    : data(std::move(data)), version(version) {}
+
+  ByteContainer data{};
+  uint32_t version{0};
+};
+
+struct ExactMatchKey : public MatchKey {
+  using MatchKey::MatchKey;
+
+  static constexpr MatchUnitType mut = MatchUnitType::EXACT;
+};
+
+struct LPMMatchKey : public MatchKey {
+  LPMMatchKey() {}
+  LPMMatchKey(ByteContainer data, int prefix_length, uint32_t version)
+    : MatchKey(data, version), prefix_length(prefix_length) {}
+
+  int prefix_length{0};
+
+  static constexpr MatchUnitType mut = MatchUnitType::LPM;
+};
+
+struct TernaryMatchKey : public MatchKey {
+  TernaryMatchKey() {}
+  TernaryMatchKey(ByteContainer data, ByteContainer mask, int priority,
+      uint32_t version)
+    : MatchKey(data, version), mask(std::move(mask)),
+    priority(priority) {}
+
+  ByteContainer mask{};
+  int priority{std::numeric_limits<decltype(priority)>::max()};
```

Comment:
The reason for this is that in the lookup function it looks for the match with the minimum priority, starting from `numeric_limits<>::max()`, this ensures that default-constructed keys will never match. I can add a comment to that effect